### PR TITLE
Setup repo to support client generation with tsp-client

### PR DESCRIPTION
**Pull request by catalinaperalta**
Adding configs necessary to generate client libraries with tsp-client.

Example of how to update emitter-package.json in the future:
```
PS C:\repos\azure-sdk-for-rust> .\eng\common\scripts\typespec\New-EmitterPackageJson.ps1 -PackageJsonPath <path to>\typespec-rust\packages\typespec-rust\package.json -OutputDirectory .\eng\
```

**_There will be a tsp-client specific command coming soon!_**

To sync the lock file with the emitter-package.json file run: `tsp-client generate-lock-file`

### CODE REVIEW

File: eng/emitter-package.json
Comment (catalinaperalta) [inline eng/emitter-package.json:9]: NOTE: I manually added this for storage, however we can remove if it's not needed for now. This should be added in the peerDependencies of the typespec-rust package for it to get carried over to emitter-package.json whenever you use the script to sync dependencies.

Comment (jhendrixMSFT) [reply eng/emitter-package.json:9]: Do all tsp packages need to be listed as peer dependencies in an emitter's `package.json`? I thought it was only the ones the emitter directly uses, but I could be wrong.

Comment (catalinaperalta) [reply eng/emitter-package.json:9]: That is the convention the emitters have been following, this allows emitters to pin supported peer dependency ranges. We used to just manually pin these deps in emitter-package.json but I believe @hallipr worked on shifting to using the peer deps in the emitter's package.json so he could share more. From a tsp-client perspective we simply install whatever deps are defined in emitter-package.json. Examples:
[typespec-python](https://github.com/Azure/autorest.python/blob/f3f6491ca1d653b584b092b07f7907237cc00c4e/packages/typespec-python/package.json#L47-L58)
[typespec-csharp](https://github.com/Azure/autorest.csharp/blob/b11bb94bc90e29694ab222e3b30cb43f167cb949/src/TypeSpec.Extension/Emitter.Csharp/package.json#L41-L53)

**Comment by heaths** [comment]
Can we pass a URL instead of local file path? Currently, we have no plans to colocate the emitter source in this repo. Can we also specify it via config?

File: .gitignore
Comment (heaths) [inline .gitignore:17]: Why is this in the repo at all? We should use a subdirectory under `$TEMP`, which is universal across platforms.

File: eng/emitter-package-lock.json
Comment (heaths) [inline eng/emitter-package-lock.json:1]: We should add this file to `.gitattributes` like we have `Cargo.lock` files already:

```gitattributes
*package-lock.json merge=binary
```

**Review by heaths — CHANGES_REQUESTED** [review]

**Comment by catalinaperalta** [comment]
> Can we pass a URL instead of local file path? Currently, we have no plans to colocate the emitter source in this repo. Can we also specify it via config?

@hallipr can you help answer this? On the tsp-client side, I can take this into consideration for the new command that will replace the pwsh script.

File: eng/emitter-package-lock.json
Comment (catalinaperalta) [reply eng/emitter-package-lock.json:1]: This wont block maintainers from updating the file?

File: .gitignore
Comment (catalinaperalta) [reply .gitignore:17]: this is current tsp-client behavior, if you really want a change for this behavior please open an issue for tsp-client here and we can discuss: https://github.com/Azure/azure-sdk-tools/issues

I dont think this should block this PR from merging though if you want to start using tsp-client to generate client libraries, it's an implementation detail of the tool.

File: eng/emitter-package-lock.json
Comment (heaths) [reply eng/emitter-package-lock.json:1]: No. It just means it can't be merged. It's treated *in PRs only* as a binary file. It's still stored in git as text so storage is efficient. But it basically doesn't allow merges because this file should *never* be merged. Same as any lock file, which is why we already have `Cargo.lock` in there like this (serves the same purpose as `packages-lock.json`).

File: eng/emitter-package.json
Comment (hallipr) [reply eng/emitter-package.json:9]: These shouldn't be specified as range dependencies.  We started pinning any floating peer dependencies with dev dependencies to reduce the risk of supply chain vulnerabilities caused by picking up a version in CI that we haven't yet verified on a dev machine.  Even though the emitter-package-lock.json will pin us to a specific version, we can't control which version we pin to without specifying the version in devDependencies.

`npm install` will automatically satisfy deep peer dependencies. This means that locally, your package.json will make no mention of the version to pick but your package-lock.json will pin to the latest available version that satisfies the different required ranges. 

Because we generate emitter-package.json and its lock file automatically in emitter CI builds, we can't use lock file pinning alone, so we came up with the heuristic of "any dependency mentioned as both a peer dependency and dev dependency must be doing peer dependency pinning, so we'll forward that pin on to emitter-package.json"

This is codified in: [eng/common/scripts/typespec/New-EmitterPackageJson.ps1
](https://github.com/Azure/azure-sdk-for-rust/blob/feature/track2/eng/common/scripts/typespec/New-EmitterPackageJson.ps1#L33)

package.json:
```json
{
  "name": "@azure-tools/typespec-foo",
  "version": "0.35.1",
  "peerDependencies": {
    "@azure-tools/typespec-azure-core": ">=0.46.0 <1.0.0",
    "@azure-tools/typespec-azure-resource-manager": ">=0.46.0 <1.0.0",
    "@typespec/compiler": ">=0.60.0 <1.0.0",
    "another-package": "2.3.4"
  },
  "devDependencies": {
    "@azure-tools/typespec-azure-core": "~0.46.0",
    "@azure-tools/typespec-azure-resource-manager": "~0.46.0",
    "@typespec/compiler": "~0.60.1",
    "something-else": "1.2.3"
  }
}
```

emitter-package.json
```json
{
  "main": "dist/src/index.js",
  "dependencies": {
    "@azure-tools/typespec-foo": "0.35.1"
  },
  "devDependencies": {
    "@azure-tools/typespec-azure-core": "~0.46.0",
    "@azure-tools/typespec-azure-resource-manager": "~0.46.0",
    "@typespec/compiler": "~0.60.1",
  }
}
```

As far as which packages should be included, you can run `npm ls -a`.  If you have any open peer dependencies, it will fail and list them.

Comment (hallipr) [reply eng/emitter-package.json:9]: see: https://github.com/Azure/azure-sdk-for-python/blob/main/eng/emitter-package.json

**Review by hallipr — CHANGES_REQUESTED** [review]

File: eng/emitter-package.json
Comment (hallipr) [reply eng/emitter-package.json:9]: I have a PR out that will make emitter-package.json peer pinning more explicit in the emitter's package.json file:
https://github.com/Azure/azure-sdk-tools/pull/9147

Rather than specifying both peerDependencies and devDependencies in the emitter's package.json you pin the dev dependency and add the package name to a new section:

```json
{
  "name": "@azure-tools/typespec-python",
  "version": "0.35.1",
  "devDependencies": {
    "@azure-tools/typespec-azure-core": "~0.46.0",
    "@azure-tools/typespec-azure-resource-manager": "~0.46.0",
    "@typespec/compiler": "~0.60.1",
    "something-else": "1.2.3"
  },
  "azure-sdk/emitter-package-json-pinning": [
    "@azure-tools/typespec-azure-core",
    "@azure-tools/typespec-azure-resource-manager",
    "@typespec/compiler"
  ]
}
```

This avoids redeclaring transitive peer dependencies, it's less heuristic and it's more declarative.

Comment (catalinaperalta) [reply eng/emitter-package.json:9]: @hallipr so is this PR blocked until the emitter's package.json is updated to this format?

Comment (catalinaperalta) [reply eng/emitter-package.json:9]: @jhendrixMSFT I see several packages that also have ranges in devDependencies: https://github.com/Azure/typespec-rust/blob/a8c4021148d50dc3f54a44f7fd3489568c1e9dba/packages/typespec-rust/package.json#L64-L68 I think we should be pinning to a specific version according to what @hallipr mentioned.

Comment (jhendrixMSFT) [reply eng/emitter-package.json:9]: I thought the comment was for `emitter-package.json`? Other emitters use the same range which is desirable in order to test emitters against dev versions.

@hallipr can we get clarification on this?

Comment (catalinaperalta) [reply eng/emitter-package.json:9]: @hallipr if this is a new requirement, can we merge as is and then update the emitter-package.json once your new script is updated?

Comment (hallipr) [reply eng/emitter-package.json:9]: `emitter-package.json` should pin all transitive peer dependency ranges as dev dependencies to the same version that was used when testing the emitter.  The only way to reliably get the same versions in package-lock.json and emitter-package-lock.json is to be specific in package.json's devDependencies and automate the creation of emitter-package.json.

If specific version pinning in devDependencies is too much of a headache, we could consider pulling the specific version from `package-lock.json` rather than `package.json`'s devDependencies when creating `emitter-package.json`

Comment (catalinaperalta) [reply eng/emitter-package.json:9]: @jhendrixMSFT thoughts on pinning specific versions in devDeps?

I see c# and python pinning to specific versions already: 
C#: https://github.com/Azure/autorest.csharp/blob/d3615842dd87f222debf00bb1ca3beac0e953942/src/TypeSpec.Extension/Emitter.Csharp/package.json#L54-L81
Python: https://github.com/Azure/autorest.python/blob/333bc8e4b352f487ea24159fde23474fc3679e20/packages/typespec-python/package.json#L67-L92

Comment (catalinaperalta) [reply eng/emitter-package.json:9]: @hallipr have you already updated you script? since last time I ran it, it still just pulled in the peerDeps...otherwise, we'll have to manually update emitter-package.json to unblock this pr

File: .gitignore
Comment (heaths) [reply .gitignore:17]: https://github.com/Azure/azure-sdk-tools/issues/9248

File: eng/emitter-package.json
Comment (hallipr) [reply eng/emitter-package.json:9]: I would manually update the emitter-package.json, but use static pinning instead of ranged.

Comment (catalinaperalta) [reply eng/emitter-package.json]: Done! Will dismiss the request for changes

**Review by jhendrixMSFT — APPROVED** [review]
I don't know enough about how this should be structured but the values for the Rust emitter look correct.

**Comment by catalinaperalta** [comment]
> I don't know enough about how this should be structured but the values for the Rust emitter look correct.

This is the final shape. The only action for you is to pin your devDependencies in the typespec-rust emitter to a specific version instead of a range (otherwise you'll have to do it manually in your emitter-package.json updates). In the future you'll follow these steps to update this file with new rust emitter releases:

1. To sync emitter-package.json with the typespec-rust package.json, run:
```
PS C:\repos\azure-sdk-for-rust> .\eng\common\scripts\typespec\New-EmitterPackageJson.ps1 -PackageJsonPath <path to>\typespec-rust\packages\typespec-rust\package.json -OutputDirectory .\eng\
```

> NOTE: I'll let you know when tsp-client has a command for this so you can stop using the pwsh script

2. To sync your emitter-package-lock.json, run:
```
tsp-client generate-lock-file
```

And of course feel free to reach out if you have any questions!

**Review by RickWinter — APPROVED** [review]

File: eng/emitter-package.json
Comment (catalinaperalta) [reply eng/emitter-package.json:7]: Yes if 0.47.1 worked with 0.1.0 we should be fine for this PR. Resolving!

**Review by weshaggard — APPROVED** [review]

**Pull request merged**
